### PR TITLE
Added a service to manage PrestaShop versions (may deprecate _PS_VERSION_)

### PR DESCRIPTION
**Pull request by michaelKaefer**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | Adds functionality to retrieve the current PS version and to compare a given version to the current PS version. In this PR the PS version is hardcoded in AppKernel (which replaces _PS_VERSION_) - In Symfony the version is hardcoded in Kernel::VERSION. A Version object can be retrieved from the service container.
| Type?         | new feature
| Category?     | CO
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | -
| How to test?  | If the PR is ok I will provide unit tests

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/9174)

### CODE REVIEW

File: src/Core/Foundation/PrestaShopVersion/Exception/InvalidVersionException.php
Comment (sarjon) [inline src/Core/Foundation/PrestaShopVersion/Exception/InvalidVersionException.php]: no need to duplicate PrestaShop name.

```diff
- PrestaShop\PrestaShop\Core\Foundation\PrestaShopVersion\Exception
+ PrestaShop\PrestaShop\Core\Foundation\Version\Exception
```

File: src/Core/Foundation/PrestaShopVersion/PrestaShopVersion.php
Comment (PierreRambaud) [inline src/Core/Foundation/PrestaShopVersion/PrestaShopVersion.php]: Hum, you can replace both throw by a simple regex
```php
preg_match('~^(?:(\d+)\.?){1,}\d+$~', $version);
```
It's invalid if it don't match `1.N`, A version can be `1.0`, `1.9.1`, `1.2.3.4`. We can go further by adding support of alpha, beta, rc, etc. Making this class a little generic could be better for the community :)

Or you can play with this regex:
`[0-9]+(?>\.[0-9a-zA-Z]+)*(-[0-9A-Za-z-]+(\.[0-9A-Za-z-]+)*)?`
=)

File: src/Core/Foundation/PrestaShopVersion/PrestaShopVersion.php
Comment (PierreRambaud) [inline src/Core/Foundation/PrestaShopVersion/PrestaShopVersion.php]: set `AppKernel::VERSION` as parameter in constructor, so everyone will be able to use this =)

File: src/PrestaShopBundle/Controller/Admin/Improve/Design/ThemeCatalogController.php
Comment (PierreRambaud) [inline src/PrestaShopBundle/Controller/Admin/Improve/Design/ThemeCatalogController.php]: It's maybe redundant:
`PrestaShop\PrestaShop\Core\Foundation\Version`

File: app/AppKernel.php
Comment (PierreRambaud) [inline app/AppKernel.php:36]: Not sure it's a good idea to register in const `MAJOR`, `MINOR` and `RELEASE` const, we can get this value from `VERSION`. We can retrieve this information from the string. Making this more generic can be useful for the community in case they also need to manage version (from dependencies for example). :)

File: src/Core/Foundation/PrestaShopVersion/PrestaShopVersion.php
Comment (PierreRambaud) [inline src/Core/Foundation/PrestaShopVersion/PrestaShopVersion.php]: Must be a parameter in constructor =)

**Comment by PierreRambaud** [comment]
I like this PR, but as I said in comments, making this class more generic will be awesome \o/
(We like to plya with class like this in ruby https://github.com/rubygems/rubygems/blob/master/lib/rubygems/version.rb )

File: src/Core/Foundation/PrestaShopVersion/PrestaShopVersion.php
Comment (michaelKaefer) [reply src/Core/Foundation/PrestaShopVersion/PrestaShopVersion.php]: @PierreRambaud I will change it to a regex. - RC and Beta releases have the same version hardcoded like the stable release as far as I checked it (e.g. "1.7.3.0."). - Are you sure that it's good to make this class generic? I would keep it as strict as possible, if the string is d+.d+.d+.d+ it is valid, otherwise not. Why do we need 1.0 etc? I saw something like that hardcoded to ControllerTest.php but is there a strong reason to do that? - The method is private by the way

File: src/Core/Foundation/PrestaShopVersion/PrestaShopVersion.php
Comment (michaelKaefer) [reply src/Core/Foundation/PrestaShopVersion/PrestaShopVersion.php]: @PierreRambaud The method is private and not meant to be used outside the class. There are quite some other methods to compare a version string to the actual PS version. Is this ok?

File: app/AppKernel.php
Comment (michaelKaefer) [reply app/AppKernel.php:36]: @PierreRambaud I did it because Symfony did it :rofl:  (https://github.com/symfony/http-kernel/blob/master/Kernel.php) I don't know why they did, I just thought there must be some good reason. Should I remove it still? - What do you mean by "manage version from dependencies"?

**Comment by michaelKaefer** [comment]
@PierreRambaud Thank you for your review! I will work on it :+1:

File: src/Core/Foundation/PrestaShopVersion/PrestaShopVersion.php
Comment (michaelKaefer) [reply src/Core/Foundation/PrestaShopVersion/PrestaShopVersion.php]: @PierreRambaud Thanks, I made a commit and it looks much better now :)

File: src/Core/Foundation/PrestaShopVersion/Exception/InvalidVersionException.php
Comment (michaelKaefer) [reply src/Core/Foundation/PrestaShopVersion/Exception/InvalidVersionException.php]: @sarjon Thank you! I wanted to make it explicit :) Anyway, I changes it and it looks better now :+1:

File: src/PrestaShopBundle/Controller/Admin/Improve/Design/ThemeCatalogController.php
Comment (michaelKaefer) [reply src/PrestaShopBundle/Controller/Admin/Improve/Design/ThemeCatalogController.php]: @PierreRambaud It's because auf PSR 4 autoloading. If there is a directory Version with a class Version this is necessary. Or do I get something wrong? By the way I changed "PrestaShopVersion" to "Version" :+1:

File: src/Core/Foundation/PrestaShopVersion/PrestaShopVersion.php
Comment (michaelKaefer) [reply src/Core/Foundation/PrestaShopVersion/PrestaShopVersion.php]: @PierreRambaud I forget the most important: it is hard to accept "1.0" because version_compare() says it is less than "1.0.0". So it would be neceassary to handle all these cases seperatly which sounds bad. I changed it to a regex, please have a look. If a developer wants to compare "1.0" to the current PS version he has to transform it to "1.0.0.0" before - this does not seem too bad for me. Are you OK with this?

File: src/Core/Foundation/Version/Version.php
Comment (PierreRambaud) [inline src/Core/Foundation/Version/Version.php]: Watch out, as I said, not all versions have x.x.x.x. Can be x.x, x.x.x, and x.x.x.x (and sometimes, x.x.x-(alpha|beta|rc\d+) )

File: src/Core/Foundation/PrestaShopVersion/PrestaShopVersion.php
Comment (PierreRambaud) [reply src/Core/Foundation/PrestaShopVersion/PrestaShopVersion.php]: Making this class working for modules can be a good idea I think, not only for "How prestashop managed their versions", wdyt @mickaelandrieu ?

File: src/PrestaShopBundle/Controller/Admin/Improve/Design/ThemeCatalogController.php
Comment (PierreRambaud) [reply src/PrestaShopBundle/Controller/Admin/Improve/Design/ThemeCatalogController.php]: I didn't find anything about PS4 autoloader and version directory :/ https://www.php-fig.org/psr/psr-4/ . Even Symfony doesn't have `Version\Version`, can you show me where do you see that? I think you can remove `Version` directory and have `Version.php` inside `Foundation`, wdyt master @mickaelandrieu ?

File: src/Core/Foundation/Version/Version.php
Comment (michaelKaefer) [reply src/Core/Foundation/Version/Version.php]: @PierreRambaud I know, I already answered above :)

File: app/AppKernel.php
Comment (PierreRambaud) [reply app/AppKernel.php:36]: In this case, we can still have it in dependency injection.
If you have modules or themes for example, which need to manage some versions for anything like API versions, dependencies versions, modules versions or manage his own version, having a class to do it will be so useful for everyone. No need to rebuild the wheel :)

File: src/PrestaShopBundle/Resources/config/services/core/foundation.yml
Comment (PierreRambaud) [inline src/PrestaShopBundle/Resources/config/services/core/foundation.yml]: 👍

**Comment by mickaelandrieu** [comment]
minor issue:

>> The !php/const: tag to indicate dumped PHP constants is deprecated since Symfony 3.4 and will be removed in 4.0. Use the !php/const (without the colon) tag instead on line 5 at line 5 (near "- !php/const:AppKernel::VERSION").

File: app/AppKernel.php
Comment (mickaelandrieu) [inline app/AppKernel.php:35]: I already wanted to introduce this but @eternoendless was against this change because this makes the release harder to do (not that hard, but still^^).

So before continue, we need @eternoendless approval.

File: src/Adapter/Product/AdminProductDataProvider.php
Comment (mickaelandrieu) [inline src/Adapter/Product/AdminProductDataProvider.php:35]: further improvement: inject the shop version in constructor.

File: src/Adapter/Requirement/CheckMissingOrUpdatedFiles.php
Comment (mickaelandrieu) [inline src/Adapter/Requirement/CheckMissingOrUpdatedFiles.php:29]: same here, this should be injected in constructor.

File: src/Core/Foundation/Version/Version.php
Comment (mickaelandrieu) [inline src/Core/Foundation/Version/Version.php]: Why 2 functions? Because I think a function shouldn't do 2 different things when we can do otherwise. More this make each function easier to test and easier to understand for your IDE (do I need a string or an integer?)

File: src/Core/Foundation/Version/Exception/InvalidVersionException.php
Comment (mickaelandrieu) [inline src/Core/Foundation/Version/Exception/InvalidVersionException.php:41]: what do you think about it @michaelKaefer?

I have the feeling this make the class useful and the code easier to read in `Version`.

File: src/PrestaShopBundle/Install/Upgrade.php
Comment (mickaelandrieu) [inline src/PrestaShopBundle/Install/Upgrade.php:322]: it is dead code by design, isn't it?

File: src/PrestaShopBundle/Install/Upgrade.php
Comment (mickaelandrieu) [inline src/PrestaShopBundle/Install/Upgrade.php:705]: it is dead code by design, isn't it?

File: src/PrestaShopBundle/Install/Upgrade.php
Comment (mickaelandrieu) [inline src/PrestaShopBundle/Install/Upgrade.php:764]: it is dead code by design, isn't it?

File: src/PrestaShopBundle/Service/DataProvider/Marketplace/ApiClient.php
Comment (mickaelandrieu) [inline src/PrestaShopBundle/Service/DataProvider/Marketplace/ApiClient.php:50]: Interesting side effect of this refactoring, this allow to use a very specific version of addons API for our tests: no more bugs because the API for the next version is not set up :-)

**Comment by mickaelandrieu** [comment]
@michaelKaefer I added a commit to refactor some parts of PrestaShop using your new service + refactoring of some parts of your work.

If you disagree with it, feel free to remove the commit ;)

Good night!

**Comment by michaelKaefer** [comment]
@mickaelandrieu Really cool! I will have a more detailed look tomorrow to learn!

File: app/AppKernel.php
Comment (michaelKaefer) [reply app/AppKernel.php:36]: @PierreRambaud Thanks, now everything is injected, which really makes much more sense and looks much better! - For the other thing: if I had to decide it I would keep it as simple as possible and not make it more generic for other uses as long as it is not needed. Also, you know that PS version is very unique because it ignores semver.

**Comment by michaelKaefer** [comment]
@mickaelandrieu 2 questions:

1. is it better to use AppKernel::VERSION or to use Version::getFullVersion()? Isn't it better to only use one of them? AppKernel::VERSION is quicker to type and faster for sure but I would keep Version::getFullVersion() because if AppKernel::VERSION changes to App\Kernel::VERSION or to _PS_VERSION_ or whatever then it is much easier to change. In Symfony 4 it is best practise to namespace the AppKernel: https://github.com/symfony/demo/blob/master/src/Kernel.php. Then it won't change anymore maybe :) Can you decide which one to use, then I will do the commits, or do we use both, sometimes this and sometimes the other? 

2. Is it ok if I make another commit to inject it as a dependency (where you already made comments that it should be better), then I will do this!

**Comment by mickaelandrieu** [comment]
> is it better to use AppKernel::VERSION or to use Version::getFullVersion()?

it's better to injection Version::getFullVersion() everywhere, but I wanted to limit the potential bugs as big parts of PrestaShop code are not covered. This is definitively something I'd like to do :)

> In Symfony 4 it is best practise to namespace the AppKernel

and we will do it once we'll migrate to Symfony 4, let's finish the migration to Symfony3 first :trollface:

**Comment by michaelKaefer** [comment]
@mickaelandrieu Ok, I will inject Version::getFullVersion() everywhere the next days. Thank you!

**Comment by michaelKaefer** [comment]
@mickaelandrieu Last question: can we autowire controllers? Something like:

class SomeController{ public function indexAction(Request $request, Version $version)) {} }

**Comment by mickaelandrieu** [comment]
@michaelKaefer yes but at this moment I have decided that we must keep yaml declaration files.

Let us time to finish to migrate Back Office, autowiring will be done later, maybe :)

**Comment by PierreRambaud** [comment]
@eternoendless  Can you have a check on this PR when you'll can? :)

**Comment by michaelKaefer** [comment]
@mickaelandrieu @PierreRambaud , thank you. I made some more commits to inject it as a dependency (not everywhere but still). What will happen now? @eternoendless , can you review or approve this PR?

File: src/PrestaShopBundle/Service/Hook/HookEvent.php
Comment (mickaelandrieu) [inline src/PrestaShopBundle/Service/Hook/HookEvent.php]: should be in the ``if`` statement

File: src/PrestaShopBundle/Service/Hook/HookEvent.php
Comment (mickaelandrieu) [inline src/PrestaShopBundle/Service/Hook/HookEvent.php]: should be in the `if` statement in l.68

Comment (michaelKaefer) [reply src/PrestaShopBundle/Service/Hook/HookEvent.php]: @mickaelandrieu Thanks and done :+1:

**Review by mickaelandrieu — APPROVED** [review]

**Comment by mickaelandrieu** [comment]
LGTM => this doesn't change the release process, we only have to change release version in only one place :)

**Comment by mickaelandrieu** [comment]
Thanks @michaelKaefer!

File: src/Core/Foundation/Version/Version.php
Comment (PierreRambaud) [inline src/Core/Foundation/Version/Version.php:34]: @mickaelandrieu The name is still redundant `Version\Version`, why version should be an exception when we don't want to do like that for other classes :/

File: src/PrestaShopBundle/Controller/Admin/Improve/Design/ThemeCatalogController.php
Comment (PierreRambaud) [inline src/PrestaShopBundle/Controller/Admin/Improve/Design/ThemeCatalogController.php:29]: Useless line

**Comment by PierreRambaud** [comment]
@mickaelandrieu You merged it too fast I think :D

**Comment by mickaelandrieu** [comment]
it is in review state for 11 days, and we - you and I - have reviewed it at least 3 times.
If it could be improved, let's do another pull request: we have until the release of 1.7.5 ;)

File: src/Core/Foundation/Version/Version.php
Comment (michaelKaefer) [reply src/Core/Foundation/Version/Version.php:34]: @PierreRambaud As I said before I think it is a necessity for PSR-4? Is that wrong? Other question: is it kind of forbidden in PrestaShop project? In Symfony this is quite common.

File: src/PrestaShopBundle/Controller/Admin/Improve/Design/ThemeCatalogController.php
Comment (michaelKaefer) [reply src/PrestaShopBundle/Controller/Admin/Improve/Design/ThemeCatalogController.php:29]: @PierreRambaud I will fix it this evening with a new PR, thank you!

File: src/Core/Foundation/Version/Version.php
Comment (sarjon) [reply src/Core/Foundation/Version/Version.php:34]: maybe `Versioning\Version` would be better?

File: src/Core/Foundation/Version/Version.php
Comment (sarjon) [inline src/Core/Foundation/Version/Version.php:77]: `$this->version` is not defined in class as `private $version;`

File: src/Core/Foundation/Version/Version.php
Comment (sarjon) [inline src/Core/Foundation/Version/Version.php:91]: this always returns `null`. You should use `$this->version` instead or rename `$this->version` to `$this->fullVersion`

File: src/Core/Foundation/Version/Version.php
Comment (sarjon) [inline src/Core/Foundation/Version/Version.php:141]: do you mean `InvalidVersionException` instead of `InvalidArgumentException`?

File: src/Core/Foundation/Version/Version.php
Comment (sarjon) [inline src/Core/Foundation/Version/Version.php:75]: wouldnt it be better to pass single full version like `'1.7.4.0'` and then parse different formats instead of passing a lot of versions? @mickaelandrieu what do you think?

**Comment by mickaelandrieu** [comment]
> this always returns null. You should use $this->version instead or rename $this->version to $this->fullVersion

> $this->version is not defined in class as private $version;

> do you mean InvalidVersionException instead of InvalidArgumentException?

@sarjon would you mind to do a pull request to do the updates?

> wouldnt it be better to pass single full version like '1.7.4.0' and then parse different formats instead of passing a lot of versions? @mickaelandrieu what do you think?

I'm not sure yet, but if you think this is better, add it to your pull request :)

> maybe Versioning\Version would be better?

I'm not really fan of `Versioning`, how about `Version\ShopVersion` instead ?

File: src/Core/Foundation/Version/Version.php
Comment (michaelKaefer) [reply src/Core/Foundation/Version/Version.php:77]: @sarjon Ups, I will fix it in the evening, thank you!

File: src/Core/Foundation/Version/Version.php
Comment (michaelKaefer) [reply src/Core/Foundation/Version/Version.php:141]: @sarjon Ups, I will fix it in the evening, thank you!

File: src/Core/Foundation/Version/Version.php
Comment (michaelKaefer) [reply src/Core/Foundation/Version/Version.php:91]: @sarjon Ups, I will fix it in the evening, thank you!

File: src/Core/Foundation/Version/Version.php
Comment (michaelKaefer) [reply src/Core/Foundation/Version/Version.php:75]: @sarjon @PierreRambaud said the same above. I just did it like Symfony does it without knowing the reason. Anyway I like it like it is and Symfony will have some goog reason. If you change it in another PR it's also ok for me of course.

**Comment by mickaelandrieu** [comment]
@michaelKaefer you need to make a new pull request as this one is merged, thanks for your help :+1:

**Comment by PierreRambaud** [comment]
> I'm not really fan of Versioning, how about Version\ShopVersion instead ?

> As I said, we maybe can make this class generic in case we use this for modules, `src/Core/Foundation/Version.php`, it can be a base about how PrestaShop use versions for core, modules, etc.

Ok to make it in `Core/Foundation` directly

**Comment by michaelKaefer** [comment]
@PierreRambaud As I said I don't think this is a good idea. The class is for PS versions only because nearly nobody else in the PHP world (modules, packages, etc.) will use the format with 4 numbers, only PS does.

And I think we need a directory - or where can we put the exception class?

**Comment by mickaelandrieu** [comment]
> And I think we need a directory - or where can we put the exception class?

in `src/Core/Foundation/Exception`? ping @PierreRambaud we need a decision here before michael starts to work on it please :)

**Comment by michaelKaefer** [comment]
@mickaelandrieu @PierreRambaud  I really don't see why Version/Version is bad. It's common to use redundant names in other well coded projects and I don't consider it as bad practise at all. And in my opinion it's much better than ShopVersion and similar inventions..

**Comment by michaelKaefer** [comment]
@sarjon @PierreRambaud @mickaelandrieu I fixed everything in a new PR (except from the Version\Version thing).

**Comment by michaelKaefer** [comment]
@PierreRambaud Just two random examples: https://github.com/Seldaek/monolog/blob/master/src/Monolog/Handler/Handler.php https://github.com/symfony/form/blob/master/Form.php

**Comment by PierreRambaud** [comment]
Yes I know, but for Symfony, it's because class is in Form tool, Version isn't a tool, it always depends in which context the class is. Same for Monolog, if you see, this class is never used directly, it's an abstract class (and there's also AbstractHandler.php, it's weird).

**Pull request merged**
